### PR TITLE
Remove metal-promise,metal-ajax and metal-struct dependencies 

### PR DESCRIPTION
**Pull request by julien**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,14 +1,37 @@
...
+
+env: PATH=~/npm/node_modules/.bin:$PATH
+
+before_install: |
+  [[ ! -x ~/npm/node_modules/.bin/npm ]] && {
+    cd ~/npm && npm install npm@^5.8.0
+    cd -
+  } || true
+  # avoids bugs around https://github.com/travis-ci/travis-ci/issues/5092
+  export PATH=~/npm/node_modules/.bin:$PATH
```
Comment (wincent) [inline .travis.yml]: I have a feeling that if you change:

```
env: PATH=~/npm/node_modules/.bin:$PATH
```

to:

```
env:
  global:
    - PATH=~/npm/node_modules/.bin:$PATH
```

then you might not need line 16 here.

File: src/app/App.js [hunk 2]
```diff
@@ -612,7 +619,7 @@ class App extends EventEmitter {
 		});
 		if (!utils.isCurrentBrowserPath(path)) {
 			if (this.isNavigationPending && this.pendingNavigate) {
-				this.pendingNavigate.thenAlways(() => this.removeScreen(path), this);
+				this.pendingNavigate.then(() => this.removeScreen(path), this);
```
Comment (wincent) [inline src/app/App.js]: To preserve the existing behavior should this be?

```javascript
this.pendingNavigate
  .then(() => this.removeScreen(path))
  .catch(error => {
    this.removeScreen(path);
    throw error;
  });
```

I haven't seen the definition of `thenAlways()` yet, but I am assuming the second `this` parameter is to set the context of the callback function? If so, it was a no-op in the old function (because we were using an arrow), and in the new version we shouldn't be passing `this` as a second param to `then()` because that is supposed to be an `onReject` handler.

File: src/app/App.js [hunk 3]
```diff
@@ -768,16 +775,16 @@ class App extends EventEmitter {
...
 	maybePreventActivate_(nextScreen) {
-		return CancellablePromise.resolve()
+		return Promise.resolve()
 			.then(() => {
 				return nextScreen.beforeActivate();
 			})
 			.then(prevent => {
 				if (prevent) {
-					this.pendingNavigate = CancellablePromise.reject(new CancellablePromise.CancellationError('Cancelled by next screen'));
+					this.pendingNavigate = Promise.reject(new Error('Cancelled by next screen'));
```
Comment (wincent) [inline src/app/App.js]: So how were these promises previously used. Were any callers explicitly `catch()`-ing `CancellablePromise.CancellationError` instances? Just wondering because we may want to reject with something more specific than just an `Error` (because ultimately this is not actually an "Error" condition, is it, but our use of `Error` makes it look like one).

File: src/app/App.js [hunk 4]
```diff
@@ -1310,7 +1322,7 @@ class App extends EventEmitter {
 			}
 		};
 
-		return new CancellablePromise((resolve) => sync() & async.nextTick(() => sync() & resolve()));
+		return new Promise((resolve) => sync() & async.nextTick(() => sync() & resolve()));
```
Comment (wincent) [inline src/app/App.js]: Mentioned this to you offline because this line is very non-obvious. You said you have to loop back and deal with `async` here.

File: src/screen/RequestScreen.js [hunk 5]
```diff
@@ -195,50 +192,72 @@ class RequestScreen extends Screen {
...
+					return resp.text();
+				})
+				.then(text => {
+					if (httpMethod === RequestScreen.GET && this.isCacheable()) {
+						this.addCache(text);
+					}
+					return text;
+				}).catch((err) => {
+					throw err;
+				}),
```
Comment (wincent) [inline src/screen/RequestScreen.js]: Does this `catch` have any effect? I would expect the code to behave the same if you delete it, unless I'm mistaken.

File: src/screen/RequestScreen.js [hunk 6]
```diff
@@ -266,17 +285,13 @@ class RequestScreen extends Screen {
...
-		return request.getResponseHeader(RequestScreen.X_REQUEST_URL_HEADER);
+		return request.headers.get(RequestScreen.X_REQUEST_URL_HEADER);
 	}
 
 	/**
-	 * This function set attribute data-safari-temp-disabled to 
+	 * This function set attribute data-safari-temp-disabled to
 	 * true and set disable attribute of an input type="file" tag
-	 * is used as a polyfill for iOS 11.3 Safari / macOS Safari 11.1 
+	 * is used as a polyfill for iOS 11.3 Safari / macOS Safari 11.1
```
Comment (wincent) [inline src/screen/RequestScreen.js]: set → sets

on lines 292 and 294?

File: src/surface/Surface.js [hunk 7]
```diff
@@ -207,7 +206,7 @@ class Surface extends Disposable {
 			to = this.defaultChild;
 		}
 		this.activeChild = to;
-		return this.transition(from, to).thenAlways(() => {
+		return this.transition(from, to).then(() => {
```
Comment (wincent) [inline src/surface/Surface.js]: Is this another one that needs a `catch()` to match the old behavior?

File: test/screen/RequestScreen.js [hunk 8]
```diff
@@ -170,63 +189,82 @@ describe('RequestScreen', function() {
...
+		fetchStub.returns(Promise.resolve(
+			new Response('', {status: 200})
+		));
+
 		var screen = new RequestScreen();
 		screen.load('/url').then(() => {
-			assert.strictEqual(RequestScreen.POST, screen.getRequest().method);
-			assert.ok(screen.getRequest().requestBody instanceof FormData);
+			assert.strictEqual(RequestScreen.POST, screen.getRequest().method.toLowerCase());
+			// assert.ok(res instanceof FormData);
```
Comment (wincent) [inline test/screen/RequestScreen.js]: Commented out assertion.

File: test/screen/RequestScreen.js [hunk 9]
```diff
@@ -170,63 +189,82 @@ describe('RequestScreen', function() {
...
 				assert.ok(error.requestError);
 				done();
 			});
-		this.requests[0].error();
 	});
 
 	it('should form navigate force post method and request body wrapped in FormData', (done) => {
 		globals.capturedFormElement = globals.document.createElement('form');
+
+		let formData = new FormData(globals.capturedFormElement);
```
Comment (wincent) [inline test/screen/RequestScreen.js]: This seems to be unused? Or maybe the GitHub UI is blinding me...

File: .travis.yml (hunk 1 above)
Comment (julien) [reply .travis.yml]: Thanks I'll try that

File: src/app/App.js (hunk 2 above)
Comment (julien) [reply src/app/App.js]: Yes you're right, good catch.

File: src/app/App.js (hunk 4 above)
Comment (julien) [reply src/app/App.js]: Yes, so I looked into this and I'm a bit confused mostly because of how [`async`](https://github.com/metal/metal.js/blob/17e5fb69956a5c14d367addc1af0898a89cff42e/packages/metal/src/async/async.js) is written: it uses `MessageChannel` (which is something I've never used) and then falls back to whatever is available, in somecases that just means `setTimeout(fn, 0)`. I'm surprised we've been using this for so long given the fact that we didn't even write it but  `¯\_(ツ)_/¯`.

I re-wrote this bit: 

```javacript
return new Promise((resolve) => sync() & async.nextTick(() => sync() & resolve()));
```

as this:

```javacript
return new Promise((resolve) => {
    return sync() & (setTimeout(() => {
        return sync() & resolve();
    }), 0);
});
```

And it _looks_ like it's working in the same way, but I'm not 100% sure, so just for the sake of safety I don't want to remove `async` just yet, although I definetly think that we should look for an alternative (maintaining or making changes to this project is not an easy task).

File: src/app/App.js (hunk 3 above)
Comment (julien) [reply src/app/App.js]: I found these [examples](https://github.com/liferay/senna.js/search?q=catch&unscoped_q=catch).

File: src/screen/RequestScreen.js (hunk 5 above)
Comment (julien) [reply src/screen/RequestScreen.js]: Yes that seems to be the case - not needed.

File: src/surface/Surface.js (hunk 7 above)
Comment (julien) [reply src/surface/Surface.js]: Yes, correct. Thanks

File: test/screen/RequestScreen.js (hunk 9 above)
Comment (julien) [reply test/screen/RequestScreen.js]: It was unused yes.

File: src/app/App.js [hunk 10]
```diff
@@ -1278,13 +1298,10 @@ class App extends EventEmitter {
 	}
 
 	/**
-	 * Cancels pending navigate with <code>Cancel pending navigation</code> error.
+	 * Resets pending navigate.
 	 * @protected
 	 */
 	stopPendingNavigate_() {
-		if (this.pendingNavigate) {
-			this.pendingNavigate.cancel('Cancel pending navigation');
```
Comment (matuzalemsteles) [inline src/app/App.js]: This probably should not keep? Since the behavior was to issue an error message and "cancel" the promise...

File: src/app/App.js (hunk 4 above)
Comment (wincent) [reply src/app/App.js]: I don't think it makes sense to use the [bitwise AND operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#Bitwise_AND) (`&`) here.  That to me looks like a mistake in the original version of the code (perhaps the author meant to write `&&`); either that or it is an obfuscated attempt at doing something like what [the comma operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator) does.

In any case, `setTimeout` returns a timer ID, `resolve()` returns `undefined`, and it doesn't seem to make any sense to `&` those with whatever `sync()` returns (I don't know what that is). Likewise, returning a value from a `setTimeout()` callback makes no sense, because there is no "caller" to evaluate the returned value.

In other words, I suspect this could just be:

```javascript
return new Promise(resolve => {
  sync();
  setTimeout(() => {
    sync();
    resolve();
  }, 0);
});
```

Which seems to match the intent expressed in the commit (c2f5b618e5f555bc2eab79e99c7846dea97a105f) that added the original line.

Comment (wincent) [reply src/app/App.js]: (Edited to fix commit link.)

File: src/app/App.js (hunk 3 above)
Comment (wincent) [reply src/app/App.js]: What I meant was, "who is calling `maybePreventActivate_()` and what are they doing with thrown errors?"

- [Here is `maybePreventActivate_()`](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L785).
- [This](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L812) is where you construct your `Error` object.
- [This](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L423) is the sole call-site of `maybePreventActivate_()`.
- And [this](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L452) is the line where it will eventually receive the `Error`.
- That calls [`handleNavigateError_()`](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L613).
- Which in turn does a [`console.log` with a stack trace](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L614) and [emits a "navigationError" event](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L615) which in theory anybody could be listening to.
- `this.isNavigationPending` will be `false`, so we'll never get into [this first conditional branch](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L621), but depending on the value of `utils.isCurrentBrowserPath(path)` we still might get down to [this call to `this.removeScreen(path)`](https://github.com/liferay/senna.js/blob/257f815c83d865f4da54130df04ca41557307698/src/app/App.js#L629); I haven't looked at the definition of `isCurrentBrowserPath()`.

So at least some of that seems inappropriate for what is not actually an error condition (eg. logging a stack trace seems overkill, and emitting a "navigationError" as well) — this is just an ordinary consequence of the `beforeActivate` lifecycle method resolving to `true`.

File: src/app/App.js (hunk 4 above)
Comment (julien) [reply src/app/App.js]: Thanks for the details @wincent, that made it clearer.

File: src/app/App.js (hunk 3 above)
Comment (julien) [reply src/app/App.js]: @wincent when you have a change, please tell me what you think about the updated code. Thanks ;)

**Comment by julien** [comment]
Hey @matuzalemsteles thanks for the feedback.
Concerning IE11 yes and no :
That the unit tests all pass on IE11 and the examples don't seem broken.
Our current plan is to cut a release once this is merged and test the integration in DXP.

**Comment by matuzalemsteles** [comment]
@julien that's great to hear it. Thank you for the informations.

File: test/app/App.js [hunk 11]
```diff
@@ -867,7 +820,7 @@ describe('App', function() {
 		});
 	});
 
-	it('should restore scroll position on navigate back', (done) => {
+	xit('should restore scroll position on navigate back', (done) => {
```
Comment (wincent) [inline test/app/App.js]: Why is this one disabled? Can we either fix, or delete, or scope to run not on IE etc? If none of the above applies, a comment indicating why would be helpful.

File: test/app/App.js [hunk 12]
```diff
@@ -1732,8 +1659,12 @@ describe('App', function() {
 	});
 
 
-	it('should navigate cancelling navigation to multiple paths after navigation is scheduled to keep only the last one', (done) => {
-		const app = this.app = new App();
+	xit('should navigate cancelling navigation to multiple paths after navigation is scheduled to keep only the last one', (done) => {
```
Comment (wincent) [inline test/app/App.js]: Another `xit`.

File: test/app/App.js [hunk 13]
```diff
@@ -1769,15 +1698,19 @@ describe('App', function() {
...
+				assert.ok(!this.app.scheduledNavigationQueue.length);
 				assert.strictEqual(globals.window.location.pathname, '/path3');
 				done();
 			}
 		});
 	});
 
 
-	it('should navigate cancelling navigation to multiple paths when navigation strategy is setted up to be immediate', (done) => {
+	xit('should navigate cancelling navigation to multiple paths when navigation strategy is setted up to be immediate', (done) => {
```
Comment (wincent) [inline test/app/App.js]: Another `xit`.

File: test/screen/HtmlScreen.js [hunk 14]
```diff
@@ -4,24 +4,22 @@ import dom from 'metal-dom';
...
-		};
 		// Prevent log messages from showing up in test output.
 		sinon.stub(console, 'log');
+		fetchStub = sinon.stub(window, 'fetch');
 	});
 
 	afterEach(() => {
-		this.xhr.restore();
 		console.log.restore();
+		fetchStub.restore();
```
Comment (wincent) [inline test/screen/HtmlScreen.js]: Any reason why the pattern used for `console.log` wouldn't work here in the same way? ie.

```javascript
sinon.stub(window, 'fetch'); // before

window.fetch.restore(); //after
```

(Presuming that you can later on call `window.fetch.returns(...)` etc. I don't know if that works or not or whether you need to save a reference to the value returned by `sinon.stub()`)

File: test/app/App.js (hunk 11 above)
Comment (julien) [reply test/app/App.js]: Ok. I think this one was disabled because it was acting "flaky" at some point, but I re-enabled those 3 disabled tests and am waiting for Travis to see how this will behave, dependending on the result, I'll see what our best option is.

File: test/screen/HtmlScreen.js (hunk 14 above)
Comment (julien) [reply test/screen/HtmlScreen.js]: Thanks. Using the same pattern is a much better idea. Using `window.fetch.returns(...)` does exactly what we want so we don't need to save a reference to the value returned by `sinon.stub()`

File: src/app/App.js [hunk 15]
```diff
@@ -445,21 +443,29 @@ class App extends EventEmitter {
 			.then(() => this.syncScrollPositionSyncThenAsync_())
 			.then(() => this.finalizeNavigate_(path, nextScreen))
 			.then(() => this.maybeOverloadBeforeUnload_())
+			.then(() => {
+				this.doNavigateEnd_();
+			})
 			.catch((reason) => {
 				this.isNavigationPending = false;
 				this.handleNavigateError_(path, nextScreen, reason);
+				this.doNavigateEnd_();
```
[the remark this answers is not in the document]
Comment (julien) [reply src/app/App.js]: Thanks. I agree that it seems better to declare these as "inner" functions to avoid polluting the public API.

**Comment by julien** [comment]
So since _Travis_ is happy with these latest changes, I'll merge this in to be able to test in portal and the next dependency that will be removed is `metal-uri`. 
Thanks again @wincent and @matuzalemsteles for the review. 👍
